### DISCUSSION TITLE
Why is onload-script no longer supported?

### DISCUSSION THREAD

**Discussion by danielbayley** [npm/feedback#71] [General]
As mentioned [here](https://github.com/npm/cli/blob/latest/CHANGELOG.md#programmatic-usage) in the `CHANGELOG`…

> The `onload-script` option is no longer supported.

Why was this removed, or was this kind of functionality replaced with something else? It was useful for example to be able to write `package.json` as `package.yml`, amongst other things, but now it is broken…

What was the reasoning behind this, and is there some kind of replacement mechanism for things like this?

**Comment by darcyclarke** [comment]
Hey @danielbayley! 

The discussion/decision to remove support for the `onload-script` config actually predates my involvement in the project (~mid 2018) & was primarily discussed internally (apologize that I can't link refs); AFAICT, the idea was/is to clean up the API, avoid hairy edge cases & reduce the risk of security vulnerabilities (ie. loading an arbitrary node module into the environment when you run `npm` seemed to cause more headaches then unlock potential). I'd also say that this config was primarily useful for programmatic usage which we [haven't really ever supported](https://github.com/npm/npm/issues/8283#issuecomment-103718763) & don't recommend trying to moving forward. The team waited until v7 to clean this up since we realized it was/is a breaking change.

All that said, many of `npm`'s internal dependencies can still be used/required programmaticly if you're looking to build off of those existing capabilities (note/ref: most of our dependency resolution/logic/management was moved into a new project called [`Arborist`](https://github.com/npm/arborist) - which has it's own API).

If you want to run any command/script before you execute `npm`, I'd recommend either defining [`pre` scripts](https://docs.npmjs.com/cli/v7/using-npm/scripts) or chaining with a call to [`npm exec` / `npx`](https://docs.npmjs.com/cli/v7/commands/npx) (ex. you could run or alias something like: `npx <pkg> && npm ...`). I know it's not exactly the same but I think it should support your usecase of doing a transform of a `package.yml` into a `package.json`.

**Reply by MylesBorins** [reply]
@danielbayley does this answer your question? Do you have additional requests?

**Reply by bearfriend** [reply]
The removal of `onload-script` is made even more difficult by the fact that v7 also moved the `preinstall` script to run _after_ dependencies are installed with no mention of this as a breaking change. Is there any plan to address this in some fashion? There are multiple discussions/issues/PRs that all seem to go nowhere. Please see my [recent PR](https://github.com/npm/cli/pull/5264) for a collection of those.
